### PR TITLE
CBMC proof for aws_cryptosdk_dec_materials_destroy

### PR DESCRIPTION
**Pull request by danielsn**
*Issue #, if available:*
N/A.

*Description of changes:*
- Adds a proof harness for `aws_cryptosdk_dec_materials_destroy` function;
- Adds preconditions in `aws_cryptosdk_dec_materials_destroy` function.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/Makefile [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+# if Makefile.local exists, use it. This provides a way to override the defaults
+sinclude ../Makefile.local
+#otherwise, use the default values
+include ../Makefile.local_default
+
+include ../Makefile.string
+include ../Makefile.aws_array_list
+
+#########
+# Local vars
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/Makefile:24]: Why 1? Expected runtime?

File: .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/aws_cryptosdk_dec_materials_destroy_harness.c [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        if (materials->signctx) {
+            ensure_sig_ctx_has_allocated_members(materials->signctx);
+            __CPROVER_assume(aws_cryptosdk_sig_ctx_is_valid_cbmc(materials->signctx));
+        }
+
+        // Set up the unencrypted_data_key
+        __CPROVER_assume(aws_byte_buf_is_bounded(&materials->unencrypted_data_key, MAX_NUM_ITEMS));
+        ensure_byte_buf_has_allocated_buffer_member(&materials->unencrypted_data_key);
+        __CPROVER_assume(aws_allocator_is_valid(materials->alloc));
+        __CPROVER_assume(aws_allocator_is_valid(materials->alloc));
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/aws_cryptosdk_dec_materials_destroy_harness.c:45]: Duplicated, remove one.

Comment (feliperodri) [reply .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/aws_cryptosdk_dec_materials_destroy_harness.c:45]: Actually, also move it right after `materials->alloc = can_fail_allocator();`

File: .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/aws_cryptosdk_dec_materials_destroy_harness.c [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        // Set up the keyring trace
+        __CPROVER_assume(aws_array_list_is_bounded(
+            &materials->keyring_trace, MAX_NUM_ITEMS, sizeof(struct aws_cryptosdk_keyring_trace_record)));
+        __CPROVER_assume(materials->keyring_trace.item_size == sizeof(struct aws_cryptosdk_keyring_trace_record));
+        ensure_array_list_has_allocated_data_member(&materials->keyring_trace);
+        __CPROVER_assume(aws_array_list_is_valid(&materials->keyring_trace));
+        ensure_trace_has_allocated_records(&materials->keyring_trace, MAX_STRING_LEN);
+        __CPROVER_assume(aws_cryptosdk_keyring_trace_is_valid(&materials->keyring_trace));
+
+        __CPROVER_assume(aws_cryptosdk_dec_materials_is_valid(materials));
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/aws_cryptosdk_dec_materials_destroy_harness.c:56]: Awesome! Not for this PR, so create an issue: We should create predicates to follow the same style of all other proofs, i.e., `ensure_ aws_cryptosdk_dec_materials_is_bounded()`, `ensure_ aws_cryptosdk_dec_materials_has_allocated_members()`.

File: .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/Makefile (hunk 1 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/Makefile:24]: done

File: .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/aws_cryptosdk_dec_materials_destroy_harness.c (hunk 2 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/aws_cryptosdk_dec_materials_destroy_harness.c:45]: done

File: .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/aws_cryptosdk_dec_materials_destroy_harness.c (hunk 3 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_dec_materials_destroy/aws_cryptosdk_dec_materials_destroy_harness.c]: https://github.com/aws/aws-encryption-sdk-c/issues/527

**Pull request merged**